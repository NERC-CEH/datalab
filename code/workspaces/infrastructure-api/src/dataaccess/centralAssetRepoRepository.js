import database from '../config/database';
import centralAssetMetadataModel from '../models/centralAssetMetadata.model';

const TYPE = 'centralAssetMetadata';

const CentralAssetMetadata = () => database.getModel(centralAssetMetadataModel.modelName);

async function createMetadata(metadata) {
  const [document] = await CentralAssetMetadata().create([metadata], { setDefaultsOnInsert: true });
  return document;
}

async function updateMetadata({ assetId, ownerUserIds, visible, projectKeys }) {
  const document = await CentralAssetMetadata().findOneAndUpdate(
    { assetId },
    { ownerUserIds, visible, projectKeys },
    { new: true },
  );
  return document;
}

async function listMetadata() {
  return CentralAssetMetadata().find().exec();
}

async function metadataAvailableToProject(projectKey) {
  return CentralAssetMetadata()
    .find()
    .or([
      { visible: 'PUBLIC' },
      { visible: 'BY_PROJECT', projectKeys: { $elemMatch: { $eq: projectKey } } },
    ])
    .exec();
}

async function getMetadataWithIds(ids) {
  return CentralAssetMetadata()
    .find()
    .where('assetId')
    .in(ids)
    .exec();
}

async function metadataExists(metadata) {
  const checkingFunctions = [
    metadataWithNameVersionCombinationExists,
    metadataWithFileLocationExists,
    metadataWithMasterUrlVersionCombinationExists,
  ];
  const results = await Promise.all(checkingFunctions.map(fn => fn(metadata)));
  const conflicts = results.flatMap(result => result.conflicts);
  return { conflicts, message: conflicts.join(' ') };
}

async function metadataWithNameVersionCombinationExists({ name, version }) {
  const conflicts = [];
  if (name && version && await CentralAssetMetadata().exists({ name, version })) {
    conflicts.push(`Metadata already exists with 'name:version' combination '${name}:${version}'.`);
  }
  return { conflicts };
}

async function metadataWithFileLocationExists({ fileLocation }) {
  const conflicts = [];
  if (fileLocation && await CentralAssetMetadata().exists({ fileLocation })) {
    conflicts.push(`Metadata for asset with fileLocation '${fileLocation}' already exists.`);
  }
  return { conflicts };
}

async function metadataWithMasterUrlVersionCombinationExists({ masterUrl, version }) {
  const conflicts = [];
  if (masterUrl && await CentralAssetMetadata().exists({ masterUrl, version })) {
    conflicts.push(`Metadata for asset with 'masterUrl:version' combination '${masterUrl}:${version}' already exists.`);
  }
  return { conflicts };
}

async function assetIdExists(assetId) {
  const exists = await CentralAssetMetadata().exists({ assetId });
  return exists;
}

async function setLastAddedDateToNow(assetIds) {
  return CentralAssetMetadata()
    .updateMany({ assetId: { $in: assetIds } }, { lastAddedDate: Date.now() })
    .exec();
}

async function deleteProject(projectKey) {
  // a project is being deleted, remove the projectKey from all assets
  return CentralAssetMetadata()
    .updateMany({}, { $pull: { projectKeys: { $in: [projectKey] } } })
    .exec();
}

export default {
  createMetadata,
  updateMetadata,
  listMetadata,
  getMetadataWithIds,
  metadataAvailableToProject,
  metadataExists,
  assetIdExists,
  setLastAddedDateToNow,
  deleteProject,
  TYPE,
};
