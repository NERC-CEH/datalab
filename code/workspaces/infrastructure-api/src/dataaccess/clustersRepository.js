import database from '../config/database';
import clusterModel from '../models/cluster.model';

const ClusterModel = () => database.getModel(clusterModel.modelName);

async function createCluster(cluster) {
  const [document] = await ClusterModel().create([cluster], { setDefaultsOnInsert: true });
  return document;
}

async function deleteCluster({ projectKey, type, name }) {
  const result = await ClusterModel()
    .find({ projectKey })
    .find({ type })
    .remove({ name })
    .exec();
  return result;
}

async function clusterExists({ projectKey, name, displayName }) {
  if (projectKey && name && await ClusterModel().exists({ projectKey, name })) {
    return (`Cluster already exists with name of '${name}' in project with key '${projectKey}'.`);
  }
  if (projectKey && displayName && await ClusterModel().exists({ projectKey, displayName })) {
    return (`Cluster already exists with display name of '${displayName}' in project with key '${projectKey}'.`);
  }
  return null;
}

async function listByProject(projectKey) {
  const documents = await ClusterModel().find({ projectKey }).exec();
  return documents;
}

async function getAll() {
  return ClusterModel().find().exec();
}

export default {
  createCluster,
  deleteCluster,
  clusterExists,
  listByProject,
  getAll,
};
