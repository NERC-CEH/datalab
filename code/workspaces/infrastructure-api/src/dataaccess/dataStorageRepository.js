import database from '../config/database';
import { DELETED } from '../models/dataStorage.model';
import config from '../config/config';

function DataStorage() {
  return database.getModel('DataStorage');
}

function getAllActive({ sub }) {
  // Exclude records tagged as deleted
  const query = filterByUser(sub, { status: { $ne: DELETED } });
  return DataStorage().find(query).exec();
}

function getAllByName(_, name) { // user, name
  // For all users
  return DataStorage().findOne({ name }).exec();
}

function getById({ sub }, id) {
  const query = filterByUser(sub, { _id: id });
  return DataStorage().findOne(query).exec();
}

function getByName({ sub }, name) {
  const query = filterByUser(sub, { name });
  return DataStorage().findOne(query).exec();
}

function createOrUpdate({ sub }, requestedDataStore) {
  const query = filterByUser(sub, { name: requestedDataStore.name });
  const dataStore = {
    ...requestedDataStore,
    ...createStorageUrls(requestedDataStore),
    ...setUsers([sub]),
  };
  return DataStorage().findOneAndUpdate(query, dataStore, { upsert: true, setDefaultsOnInsert: true }).exec();
}

function deleteByName({ sub }, name) {
  const query = filterByUser(sub, { name });
  return DataStorage().remove(query).exec();
}

function update(name, updatedValues) {
  const updateObj = {
    $set: updatedValues,
  };
  return DataStorage().findOneAndUpdate({ name }, updateObj, { upsert: false }).exec();
}

function addUsers(name, userIds) { // user, name, userId
  const updateObj = setUsers(userIds);
  return DataStorage().findOneAndUpdate({ name }, updateObj, { upsert: false, new: true }).exec();
}

function removeUsers(name, userIds) { // user, name, userId
  const updateObj = unsetUsers(userIds);
  return DataStorage().findOneAndUpdate({ name }, updateObj, { upsert: false, new: true }).exec();
}

const filterByUser = (userId, findQuery) => ({
  ...findQuery,
  users: { $elemMatch: { $eq: userId } },
});

const setUsers = userIds => ({
  $addToSet: { users: { $each: userIds } },
});

const unsetUsers = userIds => ({
  $pull: { users: { $in: userIds } },
});

const createStorageUrls = requestedDataStore => ({
  url: `https://${requestedDataStore.projectKey}-${requestedDataStore.name}.${config.get('datalabDomain')}/minio`,
  internalEndpoint: `http://minio-${requestedDataStore.name}/minio`,
});

export default { getAllActive, getAllByName, getById, getByName, createOrUpdate, deleteByName, update, addUsers, removeUsers };
