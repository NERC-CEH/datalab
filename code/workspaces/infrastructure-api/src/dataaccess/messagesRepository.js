import database from '../config/database';
import messageModel from '../models/message.model';

const MessageModel = () => database.getModel(messageModel.modelName);

async function createMessage(message) {
  const [document] = await MessageModel().create([message], { setDefaultsOnInsert: true });
  return document;
}

async function deleteMessage(id) {
  const result = await MessageModel()
    .deleteOne({ _id: id })
    .exec();
  return result;
}

async function getActiveMessages() {
  // Get the messages that have not yet expired.
  const now = Date.now();
  const documents = await MessageModel().find({ expiry: { $gt: now } }).sortByCreated('desc').exec();
  return documents;
}

async function getAll() {
  return MessageModel().find().sortByCreated('desc').exec();
}

export default {
  createMessage,
  deleteMessage,
  getActiveMessages,
  getAll,
};
