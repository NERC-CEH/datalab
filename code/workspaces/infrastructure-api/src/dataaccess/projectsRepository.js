import database from '../config/database';

const Project = () => database.getModel('Project');

async function getAll() {
  return Project().find().exec();
}

async function getByKey(projectKey) {
  return Project().findOne({ key: projectKey }).exec();
}

async function exists(projectKey) {
  return Project().exists({ key: projectKey });
}

async function create(project) {
  // needs to be in array else thinks is spread of projects and not project then options
  // as passing projects as array, create returns an array
  return Project().create([project], { setDefaultsOnInsert: true });
}

async function createOrUpdate(project) {
  return Project().findOneAndUpdate(
    { key: project.key },
    project,
    { upsert: true, setDefaultsOnInsert: true, new: true },
  );
}

async function deleteByKey(projectKey) {
  return Project().remove({ key: projectKey }).exec();
}

export default {
  getAll,
  getByKey,
  exists,
  create,
  createOrUpdate,
  deleteByKey,
};
