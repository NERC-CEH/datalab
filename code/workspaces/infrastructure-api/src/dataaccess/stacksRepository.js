import database from '../config/database';

function Stack() {
  return database.getModel('Stack');
}

function getAll(user) {
  return Stack()
    .find()
    .filterByUserSharedVisible(user)
    .exec();
}

function getAllByProject(projectKey, user) {
  return Stack()
    .find()
    .filterByProject(projectKey)
    .filterByUserSharedVisible(user)
    .exec();
}

function getAllByCategory(projectKey, user, category) {
  return Stack()
    .find({ category })
    .filterByProject(projectKey)
    .filterByUserSharedVisible(user)
    .exec();
}

function getAllByVolumeMount(projectKey, user, mount) {
  // Searches All Users
  return Stack()
    .find({ volumeMount: mount })
    .filterByProject(projectKey)
    .exec();
}

function getOneById(projectKey, user, id) {
  return Stack()
    .findOne({ _id: id })
    .filterOneByProject(projectKey)
    .filterOneByUserSharedVisible(user)
    .exec();
}

function getOneByName(projectKey, user, name) {
  // Searches All Users
  return Stack()
    .findOne({ name })
    .filterOneByProject(projectKey)
    .exec();
}

function getOneByNameUserAndCategory(projectKey, user, name, category) {
  return Stack()
    .find({ category })
    .findOne({ name })
    .filterOneByProject(projectKey)
    .filterByUser(user)
    .exec();
}

function createOrUpdate(projectKey, user, stack) {
  return Stack()
    .find()
    .filterByProject(projectKey)
    .filterByUser(user)
    .findOneAndUpdate(
      { name: stack.name },
      { ...stack, users: [user.sub] },
      { upsert: true, setDefaultsOnInsert: true },
    )
    .exec();
}

function deleteStack(projectKey, user, stack) {
  return Stack()
    .find()
    .filterByProject(projectKey)
    .filterByUser(user)
    .remove({ name: stack.name })
    .exec();
}

async function userCanDeleteStack(projectKey, user, name) {
  const stack = await getOneByName(projectKey, user, name);
  return stack.users.includes(user.sub);
}

// Function is used by kube-watcher to update stacks status. This will require an
// update when kube-watcher updated to handle projectKey.
function updateStatus(stack) {
  const { name, namespace, type, status } = stack;
  return Stack()
    .where({ name, type, projectKey: namespace })
    .updateOne({ status })
    .exec();
}

export default {
  getAll,
  getAllByProject,
  getAllByCategory,
  getOneById,
  getOneByName,
  getOneByNameUserAndCategory,
  getAllByVolumeMount,
  createOrUpdate,
  deleteStack,
  userCanDeleteStack,
  updateStatus,
};
