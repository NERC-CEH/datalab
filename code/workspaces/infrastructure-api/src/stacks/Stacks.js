import { find } from 'lodash';
import { stackTypes } from 'common';
import jupyterStack from './jupyterStack';
import rstudioStack from './rstudioStack';
import zeppelinStack from './zeppelinStack';
import siteStack from './siteStack';

// NOTE: All other stack details should come from 'common/src/config/images'
const STACKS = Object.freeze({
  JUPYTER: {
    type: stackTypes.JUPYTER,
    create: jupyterStack.createJupyterNotebook,
    delete: jupyterStack.deleteJupyterNotebook,
  },
  JUPYTERLAB: {
    type: stackTypes.JUPYTERLAB,
    create: jupyterStack.createJupyterNotebook,
    delete: jupyterStack.deleteJupyterNotebook,
  },
  ZEPPELIN: {
    type: stackTypes.ZEPPELIN,
    create: zeppelinStack.createZeppelinStack,
    delete: zeppelinStack.deleteZeppelinStack,
  },
  RSTUDIO: {
    type: stackTypes.RSTUDIO,
    create: rstudioStack.createRStudioStack,
    delete: rstudioStack.deleteRStudioStack,
  },
  RSHINY: {
    type: stackTypes.RSHINY,
    create: siteStack.createSiteStack,
    delete: siteStack.deleteSiteStack,
  },
  NBVIEWER: {
    type: stackTypes.NBVIEWER,
    create: siteStack.createSiteStack,
    delete: siteStack.deleteSiteStack,
  },
  PANEL: {
    type: stackTypes.PANEL,
    create: siteStack.createSiteStack,
    delete: siteStack.deleteSiteStack,
  },
  VOILA: {
    type: stackTypes.VOILA,
    create: siteStack.createSiteStack,
    delete: siteStack.deleteSiteStack,
  },
});

const getStack = type => find(STACKS, ['type', type]);

export default { getStack };
