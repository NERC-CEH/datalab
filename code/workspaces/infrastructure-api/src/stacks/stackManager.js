import logger from '../config/logger';
import config from '../config/config';
import Stacks from './Stacks';
import stackRepository from '../dataaccess/stacksRepository';
import { REQUESTED } from '../models/stack.model';

function createStack(user, params) {
  const { projectKey, name, type } = params;
  const stack = Stacks.getStack(type);

  if (!stack) {
    logger.error(`Could not create stack. No stack definition for type ${type}`);
    return Promise.reject({ message: `No stack definition for type ${type}` });
  }

  logger.info(`Creating new ${type} stack with name: ${name} for project: ${projectKey}`);
  return stack.create(params)
    .then(response => stackRepository.createOrUpdate(
      user, {
        ...params,
        category: stack.category,
        status: REQUESTED,
        url: `https://${projectKey}-${name}.${config.get('datalabDomain')}`,
        internalEndpoint: `http://${params.type}-${name}`,
      },
    )
      .then(() => response));
}

function deleteStack(user, params) {
  const { projectKey, name, type } = params;
  const stack = Stacks.getStack(type);

  if (!stack) {
    logger.error(`Could not delete stack. No stack definition for type ${type}`);
    return Promise.reject({ message: `No stack definition for type ${type}` });
  }

  logger.info(`Deleting stack ${name} for project: ${projectKey}`);
  return stack.delete(params)
    .then(response => stackRepository.deleteStack(user, params)
      .then(() => response));
}

export default { createStack, deleteStack };
