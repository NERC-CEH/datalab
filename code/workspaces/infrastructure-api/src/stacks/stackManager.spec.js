import Stacks from './Stacks';
import * as stackRepository from '../dataaccess/stacksRepository';
import stackManager from './stackManager';

jest.mock('./Stacks');
jest.mock('../dataaccess/stacksRepository');

const getStackMock = jest.fn();
Stacks.getStack = getStackMock;

const createOrUpdateMock = jest.fn().mockReturnValue(Promise.resolve());
const deleteStackMock = jest.fn().mockReturnValue(Promise.resolve());
stackRepository.default = {
  createOrUpdate: createOrUpdateMock,
  deleteStack: deleteStackMock,
};

const StackResolve = {
  name: 'expectedStackName',
  category: 'expectedStackCatName',
  create: vals => Promise.resolve(vals),
  delete: vals => Promise.resolve(vals),
};

const params = {
  datalabInfo: {
    domain: 'expectedDatalabDomain',
    name: 'expectedDatalabName',
  },
  projectKey: 'project',
  name: 'expectedName',
  type: 'expectedType',
  another: 'field',
};

const user = 'username';

describe('Stack Controller', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('createStack', () => {
    it('calls Stack.create with correct parameters', () => {
      getStackMock.mockReturnValue(StackResolve);

      return stackManager.createStack(user, params)
        .then(response => expect(response).toMatchSnapshot());
    });

    it('calls stack repository with correct parameters', () => {
      getStackMock.mockReturnValue(StackResolve);

      return stackManager.createStack(user, params)
        .then(() => expect(createOrUpdateMock.mock.calls).toMatchSnapshot());
    });

    it('failed Stack.create requests are caught', () => {
      getStackMock.mockReturnValue(null);

      return stackManager.createStack(user, params)
        .catch(err => expect(err).toMatchSnapshot());
    });
  });

  describe('deleteStack', () => {
    it('calls Stack.delete with correct parameters', () => {
      getStackMock.mockReturnValue(StackResolve);

      return stackManager.deleteStack(user, params)
        .then(response => expect(response).toMatchSnapshot());
    });

    it('calls stack repository with correct parameters', () => {
      getStackMock.mockReturnValue(StackResolve);

      return stackManager.deleteStack(user, params)
        .then(() => expect(deleteStackMock.mock.calls).toMatchSnapshot());
    });

    it('failed Stack.delete requests are caught', () => {
      getStackMock.mockReturnValue(null);

      return stackManager.deleteStack(user, params)
        .catch(err => expect(err).toMatchSnapshot());
    });
  });
});
