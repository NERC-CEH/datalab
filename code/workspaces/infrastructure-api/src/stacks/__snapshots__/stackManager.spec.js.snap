// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack Controller createStack calls Stack.create with correct parameters 1`] = `
Object {
  "another": "field",
  "datalabInfo": Object {
    "domain": "expectedDatalabDomain",
    "name": "expectedDatalabName",
  },
  "name": "expectedName",
  "projectKey": "project",
  "type": "expectedType",
}
`;

exports[`Stack Controller createStack calls stack repository with correct parameters 1`] = `
Array [
  Array [
    "project",
    "username",
    Object {
      "another": "field",
      "category": "expectedStackCatName",
      "datalabInfo": Object {
        "domain": "expectedDatalabDomain",
        "name": "expectedDatalabName",
      },
      "internalEndpoint": "http://expectedType-expectedName.project",
      "name": "expectedName",
      "projectKey": "project",
      "status": "requested",
      "type": "expectedType",
      "url": "https://project-expectedName.datalabs.localhost",
    },
  ],
]
`;

exports[`Stack Controller createStack failed Stack.create requests are caught 1`] = `
Object {
  "message": "No stack definition for type expectedType",
}
`;

exports[`Stack Controller deleteStack calls Stack.delete with correct parameters 1`] = `
Object {
  "another": "field",
  "datalabInfo": Object {
    "domain": "expectedDatalabDomain",
    "name": "expectedDatalabName",
  },
  "name": "expectedName",
  "projectKey": "project",
  "type": "expectedType",
}
`;

exports[`Stack Controller deleteStack calls stack repository with correct parameters 1`] = `
Array [
  Array [
    "project",
    "username",
    Object {
      "another": "field",
      "datalabInfo": Object {
        "domain": "expectedDatalabDomain",
        "name": "expectedDatalabName",
      },
      "name": "expectedName",
      "projectKey": "project",
      "type": "expectedType",
    },
  ],
]
`;

exports[`Stack Controller deleteStack failed Stack.delete requests are caught 1`] = `
Object {
  "message": "No stack definition for type expectedType",
}
`;
